Added a Poll command parameter to limit number of items in a Poll

RazCrimson opened the pull request:
Closes #150 . Take a look and tell me if you need any changes, @rgroves .

RazCrimson committed:
Added a Poll command parameter to limit number of items in a Poll

Also the corresponding data for help command has been updated.

rgroves commented on src/commands/poll.js:
```diff
@@ -2,22 +2,29 @@ const Util = require('../utils/utils.js')
 
 const MAX_POLL_TIME_MINUTES = 1440 // 24 hours in minutes (60*24)
-const MAX_ITEMS = 5
+const MAX_ITEMS = 8
```
This should be `MAX_ITEMS = 9`

Let's also remove the [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)) 2 used later in the code and create a `MIN_ITEMS` constant and use it instead. Same for the default, create `DEFAULT_ITEMS` and use it in places where the magic number 5 is in use currently.

rgroves commented on src/commands/poll.js:
```diff
@@ -2,22 +2,29 @@ const Util = require('../utils/utils.js')
     name: 'poll',
-    description: 'Generates a Poll from up to 5 random elements on the list',
-    execute: async (message, [time]) => {
+    description: 'Generates a Poll from 2 to 9 random elements on the list',
```
Let's change this to a template string and inject `MIN_ITEMS` and `MAX_ITEMS` where the literal `2` and `9` are currently. That way if the min/max is ever adjusted the message will not need to be updated manually.

rgroves commented on src/commands/poll.js:
```diff
@@ -2,22 +2,29 @@ const Util = require('../utils/utils.js')
-    description: 'Generates a Poll from up to 5 random elements on the list',
-    execute: async (message, [time]) => {
+    description: 'Generates a Poll from 2 to 9 random elements on the list',
+    execute: async (message, [time, pollLimit = 5]) => {
```
Use `DEFAULT_ITEMS` instead of literal 5 in default assignment for `pollLimit`

rgroves commented on src/commands/poll.js:
```diff
@@ -2,22 +2,29 @@ const Util = require('../utils/utils.js')
-        let maxItems = items.length < MAX_ITEMS ? items.length : MAX_ITEMS
-        let emojiList = ['1️⃣', '2️⃣', '3️⃣', '4️⃣', '5️⃣'].slice(0, maxItems)
+
+        let pollItemsCount = Math.min(
```
`Math.min` takes any number of arguments, so this could be simplified down to:
```
        let pollItemsCount = Math.min(Math.max(pollLimit, MIN_ITEMS), items.length, MAX_ITEMS)
```

rgroves commented on src/commands/poll.js:
```diff
@@ -2,22 +2,29 @@ const Util = require('../utils/utils.js')
+            Math.max(pollLimit, 2),
+            Math.min(items.length, MAX_ITEMS)
+        )
+        // List can only have items up to i = 8(1-9), Emotes will break when ranges goes beyond that.
```
Let's add a note about the max items and emotes up above the const MAX_ITEMS decalaration as well so that future devs won't think they're able to increase MAX_ITEMS beyond 9 without consequence.

rgroves commented on src/commands/help.js:
```diff
@@ -23,8 +23,10 @@ module.exports = {
+                'poll {active_time_in_minutes} [{number_of_items}]',
+                'creates a poll on the channel for 2 to 9 random elements. ' +
+                    'Uses 5 elements by default. If you want a poll with custom number of elements ' +
+                    'with no poll time limit, set the time as 0. WIP',
```
You can remove `WIP`. The codebase for ListBot is so young that, everything is somewhat of a WIP. In addition, the poll command works as is currently intended, so there's no need to identify it as a WIP.

rgroves reviewed (changes requested):
Hi @RazCrimson 👋 thanks for working on this one. Just a few changes needed per the review comments.

Also, please give the ListBot a once over manual test after making the requested changes, before issuing a PR, to ensure the behavior is as expected. Try to include boundary checks (run poll with 2 items, run a poll with 9 items, run a poll with 1 item and  a poll with 10 items to check that the logic properly caps the items to the bounds, etc.). Thanks!

RazCrimson committed:
Increased the MAX_ITEMS to 9.

Added another another const variable for MIN_ITEMS to keep the code uniform.
Renamed pollLimit to DEFAULT_ITEMS

RazCrimson committed:
Optimized Math.min() usage

RazCrimson committed:
Renamed Variables to better suit their usage and updated help command description of poll.

RazCrimson commented:
I also did a test before the PR, but I don't know how my mind mapped having 8 items to getting the right output. Sorry about that. 

Right now, I tested using a local instance with values within bound, out of bound, with and without timer. All seemed fine.

rgroves commented on src/commands/poll.js:
```diff
@@ -2,20 +2,24 @@ const Util = require('../utils/utils.js')
-    description: 'Generates a Poll from 2 to 9 random elements on the list',
-    execute: async (message, [time, pollLimit = 5]) => {
+    description: `Generates a Poll from ${MIN_ITEMS_COUNT} to ${MAX_ITEMS} random elements on the list`,
+    execute: async (message, [time, DEFAULT_ITEMS = 5]) => {
```
Hey, sorry I wasn't clear on this. I think the literal value `5` here should have been replaced with a constant. Number literals sprinkled throughout the code should be avoided as best practice. These should be named constants to make the code clearer.

So what you had before, would looked like:
```
    execute: async (message, [time, pollLimit = DEFAULT_ITEMS]) => {
```

And we'd have this declared with the other constants above (MAX_ITEMS, MIN_ITEMS, etc.):
```
const DEFAULT_ITEMS = 5
```

Note: you can change the `pollLimit` variable name to make it more descriptive and inline with the others if you want (maybe `requestedItemsCount`, etc. —up to you).

rgroves commented on src/commands/poll.js:
```diff
@@ -2,20 +2,24 @@ const Util = require('../utils/utils.js')
 const MAX_POLL_TIME_MINUTES = 1440 // 24 hours in minutes (60*24)
-const MAX_ITEMS = 8
+const MIN_ITEMS_COUNT = 2
+const MAX_ITEMS = 9
```
Please name all of these related constants similarly, if you're going to use the `_COUNT` suffix, use it for the max and default too (see other comment on line 13 about default).

So either name them all like this:
```
DEFAULT_ITEMS_COUNT
MIN_ITEMS_COUNT
MAX_ITEMS_COUNT
```
or like this:
```
DEFAULT_ITEMS
MIN_ITEMS
MAX_ITEMS
```

RazCrimson replied:
Oops, Refactor didnt get saved.

rgroves reviewed (changes requested):
A couple things to cleanup on recent changes, then this will be good to merge. Thanks!

RazCrimson committed:
Added a constant for a magic number 5.

Renamed variables to maintain uniformity

rgroves reviewed (approved):
Looks good! 🥳
